### PR TITLE
Add option "lock-config" for shell command "config:set"

### PR DESCRIPTION
**Pull request by avstudnitz**
This is similar to the "lock" switch which writes configuration
values to app/etc/env.php. The "lock-config" switch writes it to
app/etc/config.php instead which can be shared between environments. 
For consistency, the "lock" switch has been renamed to "lock-env".

### Description

In most cases, configuration settings should be shared among all development, staging and live instances of a shop. To achieve this, you can use the same mechanism as with the `--lock` switch described in http://devdocs.magento.com/guides/v2.2/config-guide/cli/config-cli-subcommands-config-mgmt-set.html.

The only difference is that the configuration values will be stored in app/etc/config.php instead of app/etc/env.php. This file should usually be included in the VCS and thus shared between instances.

To store a config setting in the file via command line, the command would be as follows:

```
bin/magento config:set --lock-config <path> <value>
```

i.e. 

```
bin/magento config:set --lock-config general/store_information/name "Sample Store"
```

### Manual testing scenarios
1. Call `bin/magento config:set --lock-config general/store_information/name "Sample Store"` from the command line
2. Verify that the app/etc/config.php file contains a section as follows:

```
[...]
  'cache_types' => 
  array (
    'config' => 1,
    [...]
  ),
  'install' => 
  array (
    'date' => 'Wed, 08 Nov 2017 11:52:21 +0000',
  ),
  'system' => 
  array (
    'default' => 
    array (
      'general' => 
      array (
        'store_information' => 
        array (
          'name' => 'Sample Store',
        ),
      ),
    ),
  ),
);
```
3. Verify that the configuration setting "Store Name" in the section "General", group "Store Information" is filled with "Sample Store" and greyed out:

grafik

### Contribution checklist
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [X] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by avstudnitz** [comment]
I am not totally happy with the naming of the option. Instead of `--share` I could think of `--lock-shared`. I think both are okay, but not perfect. I am open for suggestions.

**Comment by fooman** [comment]
Thanks @avstudnitz - read your blog post on this and an addition like this totally makes sense. Agree that naming this is tricky.

I would suggest the following: 
--lock-env (which is just an alias for --lock as to not break anything)
--lock-config

**Comment by avstudnitz** [comment]
You can find the blog post about this functionality which was mentioned by @fooman at https://www.integer-net.com/magento-2-2-configuration-via-code/.

**Comment by jalogut** [comment]
Just referencing another PR with improvements in same area:

* #12410

**Comment by avstudnitz** [comment]
I adjusted the PR accordingly (now using Virtual Classes to avoid code duplication) and updated the description. Thanks for the suggestion @fooman.

**Comment by fooman** [comment]
@avstudnitz would you mind taking a look at this failure as well https://travis-ci.org/magento/magento2/jobs/306722542#L1031 which seems related.

**Comment by avstudnitz** [comment]
@fooman Done.

File: app/code/Magento/Config/Console/Command/ConfigSet/ProcessorFacade.php [hunk 1]
```diff
@@ -96,13 +97,13 @@ public function __construct(
...
+     * @param string $lockTarget
      * @return string Processor response message
      * @throws ValidatorException If some validation is wrong
-     * @throws CouldNotSaveException If cannot save config value
-     * @throws ConfigurationMismatchException If processor can not be instantiated
+     * @internal param bool $share The share flag
      * @since 100.2.0
      */
-    public function process($path, $value, $scope, $scopeCode, $lock)
+    public function process($path, $value, $scope, $scopeCode, $lock, $lockTarget = ConfigFilePool::APP_ENV)
```
Comment (okorshenko) [inline app/code/Magento/Config/Console/Command/ConfigSet/ProcessorFacade.php]: Unfortunately this backward incompatible change. Please follow the instruction here http://devdocs.magento.com/guides/v2.0/contributor-guide/backward-compatible-development/#adding-parameters-in-public-methods

**Review by okorshenko — DISMISSED** [review]

File: app/code/Magento/Config/Console/Command/ConfigSet/ProcessorFacade.php (hunk 1 above)
Comment (avstudnitz) [reply app/code/Magento/Config/Console/Command/ConfigSet/ProcessorFacade.php]: @okorshenko Can you explain to me why adding an optional parameter at the end should be backwards incompatible? I'd like to understand that.

Comment (okorshenko) [reply app/code/Magento/Config/Console/Command/ConfigSet/ProcessorFacade.php]: Hi @avstudnitz this class has annotation `@api`. Adding new parameter to the method will break SPI clients of this class

Comment (avstudnitz) [reply app/code/Magento/Config/Console/Command/ConfigSet/ProcessorFacade.php]: Changed. If you have a better name than `processWithLockTarget` for the new method please tell me.

**Comment by fooman** [comment]
@avstudnitz Thanks for addressing @okorshenko's feedback. Would you mind taking a look at the mentioned merge conflict as well.

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=12178) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=12178) before we can accept your contribution.<br/>**8** out of **11** committers have signed the CLA.<br/><br/>:white_check_mark: gwharton<br/>:white_check_mark: ishakhsuvarov<br/>:white_check_mark: p-bystritsky<br/>:white_check_mark: serhii-balko<br/>:white_check_mark: okorshenko<br/>:white_check_mark: sidolov<br/>:white_check_mark: nmalevanec<br/>:white_check_mark: RomaKis<br/>:x: AzVo<br/>:x: andimov<br/>:x: duhon

**Comment by avstudnitz** [comment]
@fooman I resolved the merge conflict and rebased the branch. Seems that has some side effects, tell me if I should create a new clean pull request.

**Comment by okorshenko** [comment]
Hi @avstudnitz 
Something goes wrong with your branch. Could you please extract your changes and apply them to new branches based on magento/magento2 : 2.2-develop
Please create new PR once ready.
Unfortunately, we can con merge this PR as is

**Comment by avstudnitz** [comment]
Continued in https://github.com/magento/magento2/pull/13280.

**Pull request closed without merging**